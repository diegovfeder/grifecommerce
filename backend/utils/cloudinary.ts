const {
	CLOUDINARY_API_KEY,
	CLOUDINARY_API_SECRET,
	CLOUDINARY_CLOUD_NAME,
	CLOUDINARY_FOLDER,
} = process.env;

export const cloudinary = {
	apiKey: CLOUDINARY_API_KEY,
	apiSecret: CLOUDINARY_API_SECRET,
	cloudName: CLOUDINARY_CLOUD_NAME,
	folder: CLOUDINARY_FOLDER,
};
