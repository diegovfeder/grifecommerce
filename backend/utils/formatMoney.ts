const formatterOptions = {
	style: 'currency',
	currency: 'BRL',
	minimumFractionDigits: 2,
};

const formatter = new Intl.NumberFormat('pt-BR', formatterOptions);

const formatMoney = (amount = 0) => {
	return formatter.format(amount / 100);
};

export default formatMoney;
