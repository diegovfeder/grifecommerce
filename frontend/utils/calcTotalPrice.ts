import { CartItemProps } from '../@types/commonTypes';

export default function calcTotalPrice(cart: CartItemProps[]) {
	return cart.reduce((tally, cartItem) => {
		if (!cartItem.product) return tally;
		return tally + cartItem.quantity * cartItem.product.price;
	}, 0);
}
