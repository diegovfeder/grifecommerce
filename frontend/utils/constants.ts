export const TEXT_NO_DESCRIPTION = '- No description available';

export const NUM_TOTAL_PRODUCTS_PER_PAGE = 2;
