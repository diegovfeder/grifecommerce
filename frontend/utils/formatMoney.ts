const formatMoney = (amount = 0) => {
	const options = {
		style: 'currency',
		currency: 'BRL',
		minimumFractionDigits: 2,
	};

	if (amount % 100 === 0) {
		options.minimumFractionDigits = 2;
	}

	const formatter = Intl.NumberFormat('pt-BR', options);

	return formatter.format(amount / 100);
};

export default formatMoney;
