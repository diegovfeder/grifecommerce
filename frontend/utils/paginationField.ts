import { PRODUCTS_COUNT_QUERY } from '../gql/queries';

const paginationField = () => {
	return {
		keyArgs: false,
		// FIXME: Add proper types
		read(existing = [], { args, cache }: any) {
			const { take, skip } = args;
			const data = cache.readQuery({ query: PRODUCTS_COUNT_QUERY });
			const { productsCount } = data;
			const page = skip / take + 1;
			const pages = Math.ceil(productsCount / take);
			const items = existing.slice(skip, skip + take).filter(x => x);

			if (items.length && items.length !== take && page === pages) {
				return items;
			}
			if (items.length !== take) {
				return false;
			}
			if (items.length) {
				return items;
			}
			return false;
		},
		// FIXME: Add proper types
		merge(existing: any, incoming: any, { args }: any) {
			const { skip } = args;
			const merged = existing ? existing.slice(0) : [];
			for (let i = skip; i < skip + incoming.length; ++i) {
				merged[i] = incoming[i - skip];
			}
			return merged;
		},
	};
};

export default paginationField;
