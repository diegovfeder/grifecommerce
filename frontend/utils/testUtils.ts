// eslint-disable-next-line import/no-extraneous-dependencies
import casual from 'casual';
import { PRODUCTS_COUNT_QUERY } from '../gql/queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	// __typename: 'Item',
	id: '718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d',
	name: 'GRIFE Sample Pack - (Yellow)',
	description:
		'This our first sample pack for summer beats in a sunny day, when all is yellow :)',
	price: 5000,
	// user: null,
	photo: {
		// FIXME:
		// id: 'abc123',
		// altText: 'dogs are best',
		image: {
			publicUrlTransformed: '/samplepack_yellow.jpg',
		},
	},
});

const fakeUser = (rest?: any) => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...rest,
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: {
		image: `${casual.word}.jpg`,
	},
	name: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2022-12-11T20:16:13.797Z',
	user: fakeUser(),
});

const fakeCartItem = (rest?: any) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	product: fakeItem(),
	user: fakeUser(),
	...rest,
});

// FIXME: Properly type this class
// Fake LocalStorage
class LocalStorageMock {
	store: { [key: string]: any } = {};
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key: string) {
		return this.store[key] || null;
	}

	setItem(key: string, value: string) {
		this.store[key] = value.toString();
	}

	removeItem(key: string) {
		delete this.store[key];
	}
}

function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PRODUCTS_COUNT_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	];
}

export {
	makePaginationMocksFor,
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
};
