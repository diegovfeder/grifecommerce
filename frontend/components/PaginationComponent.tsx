import Head from 'next/head';
import Link from 'next/link';

import StyledPagination from './styles/StyledPagination';
import StyledPaginationContainer from './styles/StyledPaginationContainer';
import { NUM_TOTAL_PRODUCTS_PER_PAGE } from '../utils/constants';

interface PaginationComponentProps {
	page: number;
	productsCount: number;
}

const PaginationComponent = ({
	page,
	productsCount,
}: PaginationComponentProps) => {
	const pagesTotal = Math.ceil(productsCount / NUM_TOTAL_PRODUCTS_PER_PAGE);

	return (
		<StyledPaginationContainer data-test-id="pagination">
			<StyledPagination>
				<Head>
					<title>
						GRIFE | Page {page} of {pagesTotal}
					</title>
				</Head>
				<Link href={`/products/${page - 1}`}>
					<a aria-disabled={page <= 1}>← Prev</a>
				</Link>
				<p data-test-id="pageCount">
					Page {page} of {pagesTotal}
				</p>
				<p>Total Products: {productsCount}</p>
				<Link href={`/products/${page + 1}`}>
					<a aria-disabled={page >= pagesTotal}>Next →</a>
				</Link>
			</StyledPagination>
		</StyledPaginationContainer>
	);
};

export default PaginationComponent;
