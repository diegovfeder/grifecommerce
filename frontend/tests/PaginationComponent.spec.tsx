import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import { makePaginationMocksFor } from '../utils/testUtils';
import PaginationComponent from '../components/PaginationComponent';

describe('<PaginationComponent/>', () => {
	it('renders pagination, single page and one product', () => {
		const { container } = render(
			<MockedProvider mocks={makePaginationMocksFor(1)}>
				<PaginationComponent page={1} productsCount={1} />
			</MockedProvider>,
		);
		expect(container).toHaveTextContent(
			'← PrevPage 1 of 1Total Products: 1Next →',
		);
	});

	it('renders pagination, multiple products', async () => {
		const { container, debug } = render(
			<MockedProvider mocks={makePaginationMocksFor(36)}>
				<PaginationComponent page={1} productsCount={36} />
			</MockedProvider>,
		);
		await screen.findByTestId('pagination');
		expect(container).toHaveTextContent('Page 1 of 9');
		const pageCountSpan = screen.getByTestId('pageCount');
		expect(pageCountSpan).toHaveTextContent('9');
		expect(container).toMatchSnapshot();
	});

	it('disables the previous and next button when in single page', async () => {
		render(
			<MockedProvider mocks={makePaginationMocksFor(2)}>
				<PaginationComponent page={1} productsCount={2} />
			</MockedProvider>,
		);
		await screen.findByTestId('pagination');
		const prevButton = screen.getByText(/Prev/);
		const nextButton = screen.getByText(/Next/);
		expect(prevButton).toHaveAttribute('aria-disabled', 'true');
		expect(nextButton).toHaveAttribute('aria-disabled', 'true');
	});

	it('disables the next button when in the last page', async () => {
		render(
			<MockedProvider mocks={makePaginationMocksFor(11)}>
				<PaginationComponent page={3} productsCount={11} />
			</MockedProvider>,
		);
		await screen.findByTestId('pagination');
		const prevButton = screen.getByText(/Prev/);
		const nextButton = screen.getByText(/Next/);
		expect(prevButton).toHaveAttribute('aria-disabled', 'false');
		expect(nextButton).toHaveAttribute('aria-disabled', 'true');
	});

	it('enables previous and next buttons when in a middle page', async () => {
		render(
			<MockedProvider mocks={makePaginationMocksFor(24)}>
				<PaginationComponent page={4} productsCount={24} />
			</MockedProvider>,
		);
		await screen.findByTestId('pagination');
		const prevButton = screen.getByText(/Prev/);
		const nextButton = screen.getByText(/Next/);
		expect(prevButton).toHaveAttribute('aria-disabled', 'false');
		expect(nextButton).toHaveAttribute('aria-disabled', 'false');
	});
});
