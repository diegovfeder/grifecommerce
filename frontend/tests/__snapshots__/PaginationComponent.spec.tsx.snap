// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationComponent/> renders pagination, multiple products 1`] = `
<div>
  <div
    class="StyledPaginationContainer-sc-nom4q0-0 eqreFF"
    data-test-id="pagination"
  >
    <div
      class="StyledPagination-sc-1vd1tt1-0 zPWJP"
    >
      <a
        aria-disabled="true"
        href="/products/0"
        style="white-space: nowrap;"
      >
        ← Prev
      </a>
      <p
        data-test-id="pageCount"
      >
        Page 
        1
         of 
        6
      </p>
      <p>
        Total Products: 
        36
      </p>
      <a
        aria-disabled="false"
        href="/products/2"
      >
        Next →
      </a>
    </div>
  </div>
</div>
`;
